Extending linear elasticity to 3D

tsarikahin opened the pull request.

danielwolff1 commented on utils/elasticity/elasticity_utils.py:
```diff
@@ -458,4 +458,293 @@ def zero_neumann_y_mixed_formulation(x, y, X):
+    sigma_xy_y = dde.grad.jacobian(y, x, i=6, j=1)
+    #sigma_xy_z = dde.grad.jacobian(y, x, i=6, j=2)
+    
+    sigma_xz_x = dde.grad.jacobian(y, x, i=7, j=0)
+    #sigma_xz_y = dde.grad.jacobian(y, x, i=7, j=1)
+    sigma_xz_z = dde.grad.jacobian(y, x, i=7, j=2)
+    
+    #sigma_yz_x = dde.grad.jacobian(y, x, i=8, j=0)
+    sigma_yz_y = dde.grad.jacobian(y, x, i=8, j=1)
+    sigma_yz_z = dde.grad.jacobian(y, x, i=8, j=2)
```
**UPDATE**: If I see it correctly, then the first ordering also corresponds to your example, which means that you would have to modify the way you compute the derivatives here.

tsarikahin replied:
That's a very food catch! Thanks @danielwolff1. I will fix it.

danielwolff1 reviewed (commented):
Thanks for the MR and the extension to 3D. 👍 

Unfortunately I only managed to look at one file today, I will still check the rest of the files in the next days!

tsarikahin commented on utils/elasticity/elasticity_utils.py:
```diff
@@ -458,4 +458,293 @@ def zero_neumann_y_mixed_formulation(x, y, X):
+    
+    # Calculate the traction components in polar coordinates (normal and tangential)
+    Tn = Tx*nx + Ty*ny + Tz*nz
+    # For tangential components, I have some doubts, so this has to be reconsidered again.
+    Tt_x = -Ty*ny - Tz*nz
+    Tt_y = Tx*nx - Tz*nz
+    Tt_z = Tx*nx + Ty*ny
```
Yes, absolutely. This is a bit tricky actually.

tsarikahin commented on utils/elasticity/elasticity_utils.py:
```diff
@@ -458,4 +458,293 @@ def zero_neumann_y_mixed_formulation(x, y, X):
+        Tx, Ty, Tn, Tt: any
+            Traction components in cartesian (x,y) and polar coordinates (n (normal) and t (tangential))
+    '''    
+    sigma_xx =  y[:, 3:4]
+    sigma_yy =  y[:, 4:5]
+    sigma_zz =  y[:, 5:6]
+    sigma_xy =  y[:, 6:7]
+    sigma_yz =  y[:, 7:8]
+    sigma_xz =  y[:, 8:9]
```
This is the correct order:

```math
\mathbf{u} = (d_x, d_y, d_z, \sigma_{xx}, \sigma_{yy}, \sigma_{zz}, \sigma_{xy}, \underbrace{\sigma_{yz}, \sigma_{xz}})
```

I guess, it should be fine.

danielwolff1 reviewed (commented):
Sorry for letting the review come piece-wise but it's really busy at the moment and I don't really find the time to finish everything in one shot.

But I'm going to try and finish the remaining parts tomorrow!

Great work in any case! 😄

danielwolff1 commented on utils/geometry/custom_geometry.py:
```diff
@@ -5,6 +5,261 @@
+    def boundary_normal(self, x):
+        """Slice the unit normal at x for Neumann or Robin boundary conditions."""
+
+        n, _, _, uniq = self.boundary_normal_global
```
If I see it correctly, the calls to `self.boundary_normal_global` are independent of x, as you always compute the boundary normals and tangential on the whole geometry, which means that the output of this function will always be the same, no matter when it's called unless the underlying gmsh geometry changes.

Yet, you decided to call it here in `boundary_normal()` as well as `boundary_tangential_1()` and `boundary_tangential_2()` which -- if I understand correctly -- are queried multiple times with different input points during training. 

So from a performance point of view, I was wondering if it might make sense to call `self.fun_boundary_normal_global` once after initialization of the gmsh model and store the results internally and then only perform the slicing operations in `boundary_normal()`, `boundary_tangential_1()` and `boundary_tangential_2()`.

danielwolff1 commented on utils/geometry/custom_geometry.py:
```diff
@@ -5,6 +5,261 @@
+    def boundary_tangential_1(self, x):
+        """Slice the unit tangentil vector 1 at x for Neumann or Robin boundary conditions."""
+
+        _, t_1, _, uniq = self.boundary_normal_global
```
see above

danielwolff1 commented on utils/geometry/custom_geometry.py:
```diff
@@ -5,6 +5,261 @@
+    def boundary_tangential_2(self, x):
+        """Slice the unit tangentil vector 2 at x for Neumann or Robin boundary conditions."""
+
+        _, _, t_2, uniq = self.boundary_normal_global
```
see above

danielwolff1 reviewed (commented):
So, I finally managed to take a look at all files!

Nice work, the implementation looks generally very clean, I like it. I have a smaller performance-related concern, but aside from that no further remarks 👍

tsarikahin commented:
@danielwolff1 thanks a lot for the feedback. I hope I fixed all of them. 

Regarding boundary normals. Yes, you are right. But actually I thought about this very started and implemented in such a way that we do not generate boundary normals at each training iteration. 

when we generate geometry for deepxde, basically custom_geometry (GmshGeometry3D(Geometry)), I call 

```python
self.boundary_normal_global = self.fun_boundary_normal_global()
```

This only generates boundary normals just once and store it in self.boundary_normal_global as an attribute. Then in "boundary_normal" method we just do the slicing.

tsarikahin commented:
@danielwolff1 I have also some more examples, such as Lame 3D. And I thought about an implementation for determining boundary normals in case it is desired.

danielwolff1 commented on utils/postprocess/elasticity_postprocessing.py:
```diff
@@ -165,4 +165,94 @@ def solutionFieldOnMeshToVtk(X, triangles, pointData, file_path):
+        Coordinate system for polar transformation: spherical or cylindrical
+    '''
+    
+    X, offset, cell_types, elements = geom.get_mesh()
+
+    output = model.predict(X)
```
Just as a comment / idea for later:

Maybe it makes sense to extract these lines (or at least the model evaluation) from this method. I'm just thinking about cases that we discussed yesterday, where we might want to compare the PINN prediction to that of an FE simulation which was solved on a different mesh. In that case we would need to evaluate the trained model not on the geometry used for training but on the nodes of the computational mesh coming from the FEM simulation. 

So maybe we could adapt the signature to pass an `output` tensor and then the `model` parameter would not be needed anymore (if I see it correctly).

danielwolff1 replied:
**EDIT**: Okay, thinking further about this, we would still have to somehow take care of the elements, so it's not that straightforward.. Let's maybe discuss this in person :)

tsarikahin replied:
I would leave this for the time being as is. Because if there is a solution which is defined on another mesh or point cloud, such as FEM, this is another story. One has to generate own mesh, connectivity and and so on. But this can be done inside of the postprocessing method as well. Let's leave this as is for now.

danielwolff1 reviewed (commented):
Thanks a lot for the update and sorry for taking so long with the review. I only have some minor remarks, mostly regarding documentation (Kudos btw for the implementation of the coordinate transformations, I couldn't spot a mistake and I imagine it must have been quite tedious to write out the components of the matrices ^^).

As for the examples, I carefully checked the setups of `block_under_compression.py` and `block_under_shear.py` and only skimmed over the other ones. We should discuss the results at some point in detail (after your research seminar).

Thanks for putting in the hard work, nicely done! 💪

tsarikahin commented on elasticity_3d/linear_elasticity/block_under_compression.py:
Maybe not in this MR, but one needs to update Readme in any case.

tsarikahin commented on utils/elasticity/elasticity_utils.py:
```diff
@@ -458,4 +459,308 @@ def zero_neumann_y_mixed_formulation(x, y, X):
+        Stress component in the y direction
+    sigma_zz : any
+        Stress component in the z direction
+    sigma_xy : any
+        Shear stress component in the xy plane
+    sigma_xz : any
+        Shear stress component in the xz plane
+    sigma_yz : any
+        Shear stress component in the yz plane
```
https://www.javelin-tech.com/blog/2014/11/simulation-shear-stress-plots/

I guess the plane definition is correct, e.g. $\sigma_{xy}$ is actually in the xy plane. Since we do not introduce here 6 components of the shear stresses (thanks to symmetry), I don't think that we need to define the direction. It would make sense if we have both $\sigma_{xy}$ and $\sigma_{yx}$, since $\sigma_{xy}$ would be in the xy plane but in the y direction, while $\sigma_{yx}$ is still in the xy plane but in the x direction. I guess, we can keep this section as is.


![image](https://github.com/user-attachments/assets/33462b13-ce60-42aa-876a-8c0adec4d649)

tsarikahin commented:
@danielwolff1 I guess there are no further comments. If it is fine, I will go for merging.

danielwolff1 commented:
I would still add the two suggestions from my earlier reviews, but after that this can be merged for sure! 👍

tsarikahin commented on utils/elasticity/elasticity_utils.py:
```diff
@@ -458,4 +458,293 @@ def zero_neumann_y_mixed_formulation(x, y, X):
+    Returns
+    -------
+    Tx, Ty, Tz, Tn, Tt, Tn: any
+        Traction components in Cartesian (x, y, z) and polar coordinates (n and t)
```
Done

tsarikahin commented:
I guess all should be fine now. Can you approve so that we can merge it now? The two suggestions have been made but I don't know why they still appear

tsarikahin commented on utils/postprocess/elasticity_postprocessing.py:
```diff
@@ -165,4 +165,58 @@ def solutionFieldOnMeshToVtk(X, triangles, pointData, file_path):
+    model: object
+        The trained model
+    file_path: str:
+        The full file path to store the results
```
Done

danielwolff1 commented:
Sure, sure, sorry, I didn't mean to drag it out longer, thanks for incorporating everything, feel free to proceed with merging at any time! :)
